src/solvers/dg_p4est/dg_3d.jl

```diff
@@ -0,0 +1,397 @@
+                          surface_integral, dg::DGSEM)
+  # temporary buffer for projections
+  @unpack fstar_tmp1_threaded = cache
+  @unpack element_ids, node_indices = cache.mortars
+
+  size_ = (nnodes(dg), nnodes(dg), nnodes(dg))
+
+  @threaded for mortar in eachmortar(dg, cache)
+    fstar_tmp1 = fstar_tmp1_threaded[Threads.threadid()]
+
+    small_indices = node_indices[1, mortar]
+    large_indices = node_indices[2, mortar]
+
+    # Copy solution small to small
+    for pos in 1:4, j in eachnode(dg), i in eachnode(dg), v in eachvariable(equations)
+      # Use Tuple `node_indices` and `evaluate_index` to copy values
+      # from the correct face and in the correct orientation
+      cache.mortars.u[1, v, pos, i, j, mortar] = u[v, evaluate_index(small_indices, size_, 1, i, j),
+                                                      evaluate_index(small_indices, size_, 2, i, j),
+                                                      evaluate_index(small_indices, size_, 3, i, j),
+                                                      element_ids[pos, mortar]]
+    end
+
+    # Buffer to copy solution values of the large element in the correct orientation
+    # before interpolating
+    u_buffer = cache.u_threaded[Threads.threadid()]
+
+    # Copy solution of large element face to buffer in the correct orientation
+    for j in eachnode(dg), i in eachnode(dg), v in eachvariable(equations)
+      # Use Tuple `node_indices` and `evaluate_index` to copy values
+      # from the correct face and in the correct orientation
+      u_buffer[v, i, j] = u[v, evaluate_index(large_indices, size_, 1, i, j),
+                               evaluate_index(large_indices, size_, 2, i, j),
+                               evaluate_index(large_indices, size_, 3, i, j),
+                               element_ids[5, mortar]]
+    end
+
+    # Interpolate large element face data from buffer to small face locations
+    multiply_dimensionwise!(view(cache.mortars.u, 2, :, 1, :, :, mortar),
+                            mortar_l2.forward_lower,
+                            mortar_l2.forward_lower,
+                            u_buffer,
+                            fstar_tmp1)
+    multiply_dimensionwise!(view(cache.mortars.u, 2, :, 2, :, :, mortar),
+                            mortar_l2.forward_upper,
+                            mortar_l2.forward_lower,
+                            u_buffer,
+                            fstar_tmp1)
+    multiply_dimensionwise!(view(cache.mortars.u, 2, :, 3, :, :, mortar),
+                            mortar_l2.forward_lower,
+                            mortar_l2.forward_upper,
+                            u_buffer,
+                            fstar_tmp1)
+    multiply_dimensionwise!(view(cache.mortars.u, 2, :, 4, :, :, mortar),
+                            mortar_l2.forward_upper,
+                            mortar_l2.forward_upper,
+                            u_buffer,
+                            fstar_tmp1)
```

Comment:
Why is this type-unstable? It's neither in 2D, nor with `TreeMesh` in 3D.

Reply:
You don't see the type instability with `@code_warntype`, do you? I guess it's only visible in profiling results. That's #628. One difference to the `TreeMesh` is that the strides of the `view`s are less regular here. Did you play around with the memory layout?

Reply:
Not yet, I guess I'll postpone this to #642.